starknet_devnet/__init__.py

```diff
@@ -43,3 +99,522 @@ def simpler_copy(self, memo):  # pylint: disable=unused-argument
+        state=state,
+        resources_manager=resources_manager,
+        caller_address=self.caller_address,
+        contract_address=self.contract_address,
+        general_config=general_config,
+        initial_syscall_ptr=initial_syscall_ptr,
+    )
+
+    # Positional arguments are passed to *args in the 'run_from_entrypoint' function.
+    entry_points_args = [
+        self.entry_point_selector,
+        os_context,
+        len(self.calldata),
+        # Allocate and mark the segment as read-only (to mark every input array as read-only).
+        syscall_handler._allocate_segment(segments=runner, data=self.calldata),
+    ]
+
+    try:
+        runner.run_from_entrypoint(
+            entry_point.offset,
+            entry_points_args,
+            hint_locals={
+                "syscall_handler": syscall_handler,
+            },
+            static_locals={
+                "__find_element_max_size": 2**20,
+                "__squash_dict_max_size": 2**20,
+                "__keccak_max_size": 2**20,
+                "__usort_max_size": 2**20,
+                "__chained_ec_op_max_len": 1000,
+            },
+            # run_resources=tx_execution_context.run_resources,
+            verify_secure=True,
+        )
+    except VmException as exception:
+        code: ErrorCode = StarknetErrorCode.TRANSACTION_FAILED
+        if isinstance(exception.inner_exc, HintException):
+            hint_exception = exception.inner_exc
+
+            if isinstance(hint_exception.inner_exc, syscall_utils.HandlerException):
+                stark_exception = hint_exception.inner_exc.stark_exception
+                code = stark_exception.code
+                called_contract_address = (
+                    hint_exception.inner_exc.called_contract_address
+                )
+                message_prefix = (
+                    f"Error in the called contract ({hex(called_contract_address)}):\n"
+                )
+                # Override python's traceback and keep the Cairo one of the inner exception.
+                exception.notes = [message_prefix + str(stark_exception.message)]
+
+        if isinstance(exception.inner_exc, ResourcesError):
+            code = StarknetErrorCode.OUT_OF_RESOURCES
+
+        raise StarkException(code=code, message=str(exception)) from exception
+    except VmExceptionBase as exception:
+        raise StarkException(
+            code=StarknetErrorCode.TRANSACTION_FAILED, message=str(exception)
+        ) from exception
+    except SecurityError as exception:
+        raise StarkException(
+            code=StarknetErrorCode.SECURITY_ERROR, message=str(exception)
+        ) from exception
+    except HandlerException as exception:
+        raise StarkException(
+            code=exception.stark_exception.code,
+            message=str(exception.stark_exception.message),
+        ) from exception
+    except Exception as exception:
+        logger.error("Got an unexpected exception.", exc_info=True)
```

Comment:
Why is the log-and-throw pattern applied here?

Reply:
Hey Fabijan, we use this here because is part of the original implementation of the execute_entry_point module.

Reply:
Ok, found the code here: https://github.com/starkware-libs/cairo-lang/blob/de741b92657f245a50caab99cfaef093152fd8be/src/starkware/starknet/business_logic/execution/execute_entry_point.py#L238
I see one difference between the original code and yours: there's no catching of `HandlerException` (there is some handling in `VmException` handling, but I don't think that's the same).